scripts/create_perf_json.py

```diff
@@ -696,833 +707,927 @@ def extract_pebs_formula(formula: str) -> str:
-        # Info.* and topdown metrics.
-        infoname : Dict[str, str] = {}
-        # Mapping from a topdown metric name to its CPU specific formula.
-        nodes : Dict[str, str] = {}
-        # Mapping from the TMA CSV metric name to the name used in the perf json.
-        tma_metric_names : Dict[str, str] = {}
-        # Map from the column heading to the list index of that column.
-        col_heading : Dict[str, int] = {}
-        # A list of topdown levels such as 'Level1'.
-        levels : list[str] = []
-        # A list of parents of the current topdown level.
-        parents : list[str] = []
-        # Map from a parent topdown metric name to its children's names.
-        children: Dict[str, Set[str]] = collections.defaultdict(set)
-        # Map from a metric name to the metric threshold expression.
-        thresholds: Dict[str, str] = {}
-        issue_to_metrics: Dict[str, Set[str]] = collections.defaultdict(set)
-        found_key = False
-        csvf = csv.reader(csvfile)
-        for l in csvf:
-            if l[0] == 'Key':
-                found_key = True
-                for ind, name in enumerate(l):
-                    col_heading[name] = ind
-                    if name.startswith('Level'):
-                        levels.append(name)
-                if tma_cpu not in col_heading:
-                    if tma_cpu == 'ADL/RPL' and 'GRT' in col_heading:
-                        tma_cpu = 'GRT'
-                    if tma_cpu == 'MTL' and 'CMT' in col_heading:
-                        tma_cpu = 'CMT'
-                _verboseprint3(f'Columns: {col_heading}. Levels: {levels}')
-            elif not found_key:
-                continue
+    def save_form(self, name, group, form, desc, pdesc, locate, scale_unit, threshold,
+                          issues, events, infoname, aux, pmu_prefix, jo,
+                           issue_to_metrics, from_json):
+                if self.shortname == 'BDW-DE':
+                    if name in ['tma_false_sharing']:
+                        # Uncore events missing for BDW-DE, so drop.
+                        _verboseprint3(f'Dropping metric {name}')
+                        return
 
-            def field(x: str) -> str:
-                """Given the name of a column, return the value in the current line of it."""
-                return l[col_heading[x]].strip()
+                # Make 'TmaL1' group names more consistent with the 'tma_'
+                # prefix and '_group' suffix.
+                if group:
+                    if 'TmaL1' in group and 'tma_L1_group' not in group:
+                        group += ';tma_L1_group'
+                    if 'TmaL2' in group and 'tma_L2_group' not in group:
+                        group += ';tma_L2_group'
+                _verboseprint3(f'Checking metric {name}: {form}')
+                for v, _ in re.findall(r'(([A-Z_a-z0-9.]|\\-)+)', form):
+                    if v.isdigit() or re.match('\d+\.\d+', v) is not None or \
```

Comment:
Generates an invalid escape sequence. Please use raw strings `r` for regular expressions.